### DISCUSSION TITLE
Custom Yolo detection model training

### DISCUSSION THREAD

**Discussion by hardhik-08** [ultralytics/ultralytics#24178] [Q&A]
I am working with the YOLOv8m model and would like to extend it with additional custom object classes.

Specifically, I want the model to:

- Retain all the original 80 COCO classes (e.g., person, car, toothbrush, etc.)
- Add new custom classes (e.g., door, window, etc.), continuing the class indexing (e.g., 80: toothbrush, 81: door, 82: window)

I don't want to train the model from scratch with coco + custom object datasets
Any guidance or examples would be greatly appreciated.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @hardhik-08, thank you for your interest in Ultralytics 🚀! This is an automated response to help get your discussion moving, and an Ultralytics engineer will also assist soon. We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.
Since this is a custom training ❓ question, please provide as much information as possible, including dataset image examples, your dataset YAML, training logs, the exact model you are starting from, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).
If this turns out to be a 🐛 bug or unexpected behavior, please also provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.
Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Reply by glenn-jocher** [reply]
Good question — with standard `yolov8m.pt`, you can fine-tune from pretrained weights, but you can’t just append classes `81+` and keep the original COCO behavior without additional old-class supervision; otherwise the model will usually forget some COCO classes. If you want custom categories without retraining, [YOLO-World](https://docs.ultralytics.com/models/yolo-world/) is the better fit; if you need one fixed COCO+custom detector, you’ll need an incremental/distillation training setup, and the [custom trainer guide](https://docs.ultralytics.com/guides/custom-trainer/) is the right starting point.

**Comment by Y-T-G** [comment]
> I don't want to train the model from scratch with coco + custom object datasets

That's the only way to stop forgetting. You can try YOLOE with prompts if you want more classes than MSCOCO:

https://docs.ultralytics.com/models/yoloe/#text-prompt

**Reply by hardhik-08** [reply]
> > I don't want to train the model from scratch with coco + custom object datasets
> 
> That's the only way to stop forgetting. You can try YOLOE with prompts if you want more classes than MSCOCO:
> 
> https://docs.ultralytics.com/models/yoloe/#text-prompt

I did try Yoloe and Yoloworld, but they require more resources (AIPU). I have also tried your way of extending detection head, but if I need to include additional objects I need to re-train the head again.

**Reply by Y-T-G** [reply]
If you export YOLOE, it will be same speed as normal YOLO models. You can set the prompts before export and then export. The prompts get fused into the model, so during inference it's like a normal YOLO model.

**Reply by hardhik-08** [reply]
I need to run the model in axelera device, but  I think axelera does not support both YoloE and YoloWorld models.

**Reply by Y-T-G** [reply]
It should work. You can try exporting.

**Reply by tedo001** [reply]
How can YOLOE be optimized for startup products and mobile AI deployment?

**Reply by hardhik-08** [reply]
> It should work. You can try exporting.

Ok, I try this

**Reply by glenn-jocher** [reply]
Sounds good to test on Axelera, and for startup/mobile deployment the practical path is to use the smallest YOLOE checkpoint, set fixed classes once with `set_classes()`, keep `imgsz` modest, and export to an edge format like ONNX or TensorRT as shown in the [YOLOE docs](https://docs.ultralytics.com/models/yoloe/).
